### PR TITLE
[DO NOT MERGE] Stability improvements

### PR DESCRIPTION
**Pull request by timotheeg**
needs testing (especially on the dependency upgrades), but opening PR first.

@giantballofyarn

### CODE REVIEW

File: lib/master.js [hunk 1]
```diff
@@ -51,6 +54,13 @@ function Master(options) {
 		throw new Error("Heap thresholds not provided");
 	}
 
+	// Thresholds used to be provided in bytes. In 0.0.9, we prefer MB.
+	// the code below is a *cough* unreliable way to convert when needed
+	// it works under the assumption that no sane person using ipcluster would
+	// set the thresholds to 1TB RAM or more.
+	if (this.options.retire_rss   < ONE_MB) this.options.retire_rss   *= ONE_MB;
```
Comment (yangbin) [inline lib/master.js]: Seems hackish, would prefer `retire_rss_mb` and `total_maxheap`

**Comment by timotheeg** [comment]
@yangbin: as discussed, I've renamed the "official" options to introduce the ```_mb``` suffix, and supplied a more explicit option normalizer function. I'd like to keep the B<->MB converter, at least temporarily, to make it easier to use the existing configs. Please check it out again.

File: lib/master.js [hunk 2]
```diff
@@ -41,26 +41,58 @@ function isPowerOfTwo(n) {
...
+	// Thresholds used to be provided in bytes. In 0.0.9, we prefer MB.
+	// For clarity, new projects using ipcluster are expected to use the explicit new properties retire_rss_mb and total_maxheap_mb
+	// However, some projects might be passing retire_rss and totalmaxheap in MB.
+	// The code below normalizes when needed.
+	// It works based on the ^6 order of magnitudes between bytes and MB, which allow us to make 2 assumptions:
+	// if a value is supplied in bytes, it will not be less than ONE_MB (that would be insufficient for practical purposes)
+	// if a value is supplied in MB, it will not be more than ONE_MB (that would allocate 1TB RAM!)
+
+	if ('retire_rss' in options) {
+		options.total_maxheap_mb = options.retire_rss;
```
Comment (yangbin) [inline lib/master.js]: `total_maxheap_mb` typo

File: lib/master.js [hunk 3]
```diff
@@ -41,26 +41,58 @@ function isPowerOfTwo(n) {
...
+		delete options.retire_rss;
+		if (options.retire_rss_mb > ONE_MB) options.retire_rss_mb /= ONE_MB;
```
Comment (yangbin) [inline lib/master.js]: This code makes it look like we're apply the magic on `retire_rss_mb`. I think it's clearer if we do:
```javascript
if ('retire_rss' in options) {
    if (options.retire_rss > ONE_MB)
        options.retire_rss_mb = options.retire_rss / ONE_MB;
    else
        options.retire_rss_mb = options.retire_rss;

    delete options.retire_rss;
}
```

File: lib/master.js [hunk 4]
```diff
@@ -41,26 +41,58 @@ function isPowerOfTwo(n) {
...
+	}
+
+	if ('totalmaxheap' in options) {
+		options.total_maxheap_mb = options.totalmaxheap;
+		delete options.totalmaxheap;
+		if (options.total_maxheap_mb > ONE_MB) options.total_maxheap_mb /= ONE_MB;
```
Comment (yangbin) [inline lib/master.js]: ditto

File: lib/master.js (hunk 2 above)
Comment (timotheeg) [inline lib/master.js]: oh that's retarded :sweat_smile: :sob:

**Comment by timotheeg** [comment]
PR updated again!

File: lib/master.js [hunk 5]
```diff
@@ -24,10 +24,10 @@ var default_options = {
 	connect_timeout:     3  * 1000, // 3s
 	ping_timeout:        10 * 1000, // 10s
 	monitor_interval:    1  * 1000, // 1s
-	totalmaxheap:        512 * 1024 * 1024, // 512 MB
+	total_maxheap_mb:    512, // 512 MB
```
Comment (zz85) [inline lib/master.js]: if you ask me, I would prefer to keep stuff in bytes and do `512 * ONE_MB` instead.

Comment (timotheeg) [inline lib/master.js]: I see what you're saying (your comment has helped me take off a blindfold!). One of the reason I was doing this change was to allow our config to be passed directly from static config file down to ipcluster, and I didn't want to have ops have to change the configs (currently specified in MB). However This is our app's consideration. ipcluster shouldn't have to care about this and bytes do indeed give the best granularity/control.

But... with all that said, the overhead of node itself means that the rss of any node process can not be less than 1MB.... So does setting thresholds in bytes makes sense, even from ipcluster's perspective?

**Pull request closed without merging**